// @axa-fr/react-oidc/dist/core/default-component/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceWorkerNotSupported = exports.AuthenticateError = exports.Loading = exports.SessionLost = exports.CallBackSuccess = exports.Callback = exports.Authenticating = void 0;
var Authenticating_component_1 = require("./Authenticating.component");
Object.defineProperty(exports, "Authenticating", { enumerable: true, get: function () { return __importDefault(Authenticating_component_1).default; } });
var Callback_component_1 = require("./Callback.component");
Object.defineProperty(exports, "Callback", { enumerable: true, get: function () { return __importDefault(Callback_component_1).default; } });
Object.defineProperty(exports, "CallBackSuccess", { enumerable: true, get: function () { return Callback_component_1.CallBackSuccess; } });
var SessionLost_component_1 = require("./SessionLost.component");
Object.defineProperty(exports, "SessionLost", { enumerable: true, get: function () { return __importDefault(SessionLost_component_1).default; } });
var Loading_component_1 = require("./Loading.component");
Object.defineProperty(exports, "Loading", { enumerable: true, get: function () { return __importDefault(Loading_component_1).default; } });
var AuthenticateError_component_1 = require("./AuthenticateError.component");
Object.defineProperty(exports, "AuthenticateError", { enumerable: true, get: function () { return __importDefault(AuthenticateError_component_1).default; } });
var ServiceWorkerNotSupported_component_1 = require("./ServiceWorkerNotSupported.component");
Object.defineProperty(exports, "ServiceWorkerNotSupported", { enumerable: true, get: function () { return __importDefault(ServiceWorkerNotSupported_component_1).default; } });
